src/Attribute/Validator/TypeValidator.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace LastCall\DownloadsPlugin\Attribute\Validator;
+
+use Composer\Package\PackageInterface;
+use LastCall\DownloadsPlugin\Attribute\AttributeManagerInterface;
+use LastCall\DownloadsPlugin\Enum\Attribute;
+use LastCall\DownloadsPlugin\Enum\Type;
+
+class TypeValidator extends AbstractValidator
+{
+    public function __construct(
+        string $id,
+        PackageInterface $parent,
+        private AttributeManagerInterface $attributeManager
+    ) {
+        parent::__construct($id, $parent);
+    }
+
+    public function validate(mixed $value): Type
+    {
+        if (null === $value) {
+            return $this->parseUrl();
+        }
+
+        if (!\is_string($value)) {
+            $this->throwException($this->getAttribute(), sprintf('must be string, "%s" given', get_debug_type($value)));
+        }
+
+        $type = Type::tryFrom($value);
+        if (null === $type) {
+            $this->throwException($this->getAttribute(), 'is not supported');
+        }
+
+        return $type;
+    }
+
+    public function getAttribute(): string
+    {
+        return Attribute::TYPE->value;
+    }
+
+    private function parseUrl(): Type
+    {
+        $parts = parse_url($this->attributeManager->get(Attribute::URL));
+        $filename = pathinfo($parts['path'], \PATHINFO_BASENAME);
+        if (preg_match('/\.(tar\.gz|tar\.bz2)$/', $filename)) {
+            return Type::from('tar');
+        }
+        if (preg_match('/\.tar\.xz$/', $filename)) {
+            return Type::from('xz');
+        }
```

Comment:
Is there a reason why `.tar.xz` is treated separately to `.tar.gz` and `.tar.bz2`?

Also, what about the newer `.tar.zst`?

Reply:
That's a good question.

If you go to https://github.com/composer/composer/tree/main/src/Composer/Downloader and search for `extends ArchiveDownloader`, there are several results:

* `class GzipDownloader extends ArchiveDownloader`
* `class PharDownloader extends ArchiveDownloader`
* `class RarDownloader extends ArchiveDownloader`
* `class TarDownloader extends ArchiveDownloader`
* `class XzDownloader extends ArchiveDownloader`
* `class ZipDownloader extends ArchiveDownloader`

Each downloader has corresponding enum type in https://github.com/pact-foundation/composer-downloads-plugin/blob/main/src/Enum/Type.php#L5-L13

About `.tar.gz` and `.tar.bz2`, `TarDownloader` can handle both of them https://github.com/composer/composer/blob/main/src/Composer/Downloader/TarDownloader.php#L19

`.tar.xz` is handled in separate `XzDownloader` class https://github.com/composer/composer/blob/main/src/Composer/Downloader/XzDownloader.php#L20

Currently I don't think Composer support downloading and extracting `.tar.zst`. That's why this plugin also doesn't support it.

Reply:
Great! Thanks for the clear explanation :D